### PR TITLE
Chunked upload

### PR DESCRIPTION
**Pull request by vfiack**
Sharepoint limits uploads to 250MB, we need to use a specific API for big files.

### CODE REVIEW

File: src/main/java/com/nanalysis/sharepoint/SharepointClient.java [hunk 1]
```diff
@@ -155,6 +157,52 @@ public void uploadFile(String folder, String filename, byte[] data) throws IOExc
...
 
+    public void uploadBigFile(String folder, String filename, long size, InputStream input, Consumer<Double> progressCallback)
+            throws IOException, InterruptedException {
+        // create small file first, it will be overwritten later
+        uploadFile(folder, filename, "PLACEHOLDER".getBytes());
+
+        String uid = UUID.randomUUID().toString();
+        byte[] buffer = new byte[10*1024*1024]; //10MB
+        int nread;
+        int offset = 0;
```
Comment (tbparis) [inline src/main/java/com/nanalysis/sharepoint/SharepointClient.java:168]: what is nread? and offset?
nread = Number of chucks read ? Why declaring it outside the while loop? It is only used in (speed purpose?)

Comment (tbparis) [reply src/main/java/com/nanalysis/sharepoint/SharepointClient.java:168]: Oh ok, number of bytes read, which depends on the buffer. Always the buffer size until the last chunk.

File: src/main/java/com/nanalysis/sharepoint/SharepointClient.java [hunk 2]
```diff
@@ -155,6 +157,52 @@ public void uploadFile(String folder, String filename, byte[] data) throws IOExc
...
+                        siteUrl, encodePath(folder), encodePath(filename), uid));
+            } else if(offset + nread < size) {
+                // next chunck, start upload
+                uri = URI.create(String.format("%s/_api/web/GetFolderByServerRelativeUrl('%s')/Files('%s')/ContinueUpload(uploadID='%s',fileOffset=%d)",
+                        siteUrl, encodePath(folder), encodePath(filename), uid, offset));
+            } else {
+                // last chunk, finish upload
+                uri = URI.create(String.format("%s/_api/web/GetFolderByServerRelativeUrl('%s')/Files('%s')/FinishUpload(uploadID='%s',fileOffset=%d)",
+                        siteUrl, encodePath(folder), encodePath(filename), uid, offset));
+            }
```
Comment (tbparis) [inline src/main/java/com/nanalysis/sharepoint/SharepointClient.java:189]: Ok split the file and send it part by part, three cases: Start, Continue, Stop.

File: src/main/java/com/nanalysis/sharepoint/SharepointClient.java [hunk 3]
```diff
@@ -155,6 +157,52 @@ public void uploadFile(String folder, String filename, byte[] data) throws IOExc
...
+                    .header("Authorization", "Bearer " + token)
+                    .header("Accept", "application/json")
+                    .POST(HttpRequest.BodyPublishers.ofByteArray(buffer, 0, nread))
+                    .build();
+
+            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+            checkForError(response);
+
+            offset += nread;
+            progressCallback.accept(100d * offset / size);
```
Comment (tbparis) [inline src/main/java/com/nanalysis/sharepoint/SharepointClient.java:202]: Offset is then the number of bytes already sent.

File: src/test/java/com/nanalysis/sharepoint/ManualTests.java [hunk 4]
```diff
@@ -72,6 +77,15 @@ public void uploadFile() throws Exception {
...
     }
 
+    @Ignore("Manual test")
+    @Test
+    public void uploadBigFile() throws Exception {
+        File f = new File(BIG_FILE_LOCAL_PATH);
+        Consumer<Double> progress = percent -> System.out.printf("Uploading: %.2f%%%n", percent);
+        client.uploadBigFile(TEST_PATH, f.getName(), f.length(), new FileInputStream(f), progress);
+        System.out.println("File created");
+    }
```
Comment (tbparis) [inline src/test/java/com/nanalysis/sharepoint/ManualTests.java:87]: How to test that?

**Review by tbparis — COMMENTED** [review]
Not much comment on the code, but I didn't manage to test it.

**Review by tbparis — APPROVED** [review]
Ok, already tested on nmrfx side

File: src/test/java/com/nanalysis/sharepoint/ManualTests.java (hunk 4 above)
Comment (vfiack) [reply src/test/java/com/nanalysis/sharepoint/ManualTests.java:87]: Changing constants on top to target a site that you have access to (and not using OAUTH2 if using your personnal credentials) should be enough.

**Pull request merged**